src/mito2/src/sst/index/bloom_filter/creator.rs

```diff
@@ -0,0 +1,530 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+
+use common_telemetry::warn;
+use datatypes::schema::SkippingIndexType;
+use index::bloom_filter::creator::BloomFilterCreator;
+use puffin::puffin_manager::{PuffinWriter, PutOptions};
+use snafu::{ensure, ResultExt};
+use store_api::metadata::RegionMetadataRef;
+use store_api::storage::ColumnId;
+use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+use crate::error::{
+    BiErrorsSnafu, BloomFilterFinishSnafu, IndexOptionsSnafu, OperateAbortedIndexSnafu,
+    PuffinAddBlobSnafu, PushBloomFilterValueSnafu, Result,
+};
+use crate::read::Batch;
+use crate::row_converter::SortField;
+use crate::sst::file::FileId;
+use crate::sst::index::bloom_filter::INDEX_BLOB_TYPE;
+use crate::sst::index::codec::{IndexValueCodec, IndexValuesCodec};
+use crate::sst::index::intermediate::{
+    IntermediateLocation, IntermediateManager, TempFileProvider,
+};
+use crate::sst::index::puffin_manager::SstPuffinWriter;
+use crate::sst::index::statistics::{ByteCount, RowCount, Statistics};
+use crate::sst::index::TYPE_BLOOM_FILTER;
+
+/// The buffer size for the pipe used to send index data to the puffin blob.
+const PIPE_BUFFER_SIZE_FOR_SENDING_BLOB: usize = 8192;
+
+/// The indexer for the bloom filter index.
+pub struct BloomFilterIndexer {
+    /// The bloom filter creators.
+    creators: HashMap<ColumnId, BloomFilterCreator>,
+
+    /// The provider for intermediate files.
+    temp_file_provider: Arc<TempFileProvider>,
+
+    /// Codec for decoding primary keys.
+    codec: IndexValuesCodec,
+
+    /// Whether the indexing process has been aborted.
+    aborted: bool,
+
+    /// The statistics of the indexer.
+    stats: Statistics,
+
+    /// The global memory usage.
+    global_memory_usage: Arc<AtomicUsize>,
+}
+
+impl BloomFilterIndexer {
+    /// Creates a new bloom filter indexer.
+    pub fn new(
+        sst_file_id: FileId,
+        metadata: &RegionMetadataRef,
+        intermediate_manager: IntermediateManager,
+        memory_usage_threshold: Option<usize>,
+    ) -> Result<Option<Self>> {
+        let mut creators = HashMap::new();
+
+        let temp_file_provider = Arc::new(TempFileProvider::new(
+            IntermediateLocation::new(&metadata.region_id, &sst_file_id),
+            intermediate_manager,
+        ));
+        let global_memory_usage = Arc::new(AtomicUsize::new(0));
+
+        for column in &metadata.column_metadatas {
+            let options =
+                column
+                    .column_schema
+                    .skipping_index_options()
+                    .context(IndexOptionsSnafu {
+                        column_name: &column.column_schema.name,
+                    })?;
```

Comment:
non-blocking: Maybe we should store the deserialized index options in the column metadata in the future.